src/sentry/snuba/metrics/extraction.py

```diff
@@ -182,11 +201,39 @@ def is_on_demand_snuba_query(snuba_query: SnubaQuery) -> bool:
 
 
 def is_on_demand_query(dataset: Optional[Union[str, Dataset]], query: Optional[str]) -> bool:
-    """Returns ``True`` if the dataset and query combination can't be supported by standard metrics."""
+    """Returns ``True`` if the dataset is performance metrics and query contains non-standard search fields."""
+
     if not dataset or not query:
         return False
 
-    return Dataset(dataset) == Dataset.PerformanceMetrics and "transaction.duration" in query
+    return Dataset(dataset) == Dataset.PerformanceMetrics and _contains_on_demand_search_filters(
+        query
+    )
+
+
+def _contains_on_demand_search_filters(query: str) -> bool:
+    """Recursively checks if the query contains any non-standard search filters."""
+    try:
+        tokens = event_search.parse_search_query(query)
+
+        for token in tokens:
+            if _is_on_demand_search_filter(token):
+                return True
+
+        return False
+    except InvalidSearchQuery as e:
+        logger.error(f"Failed to parse search query: {query}. Error: {e}")
+        return False
+
+
+def _is_on_demand_search_filter(token: QueryToken) -> bool:
+    if isinstance(token, SearchFilter):
+        return token.key.name not in _STANDARD_METRIC_FIELDS
+
+    if isinstance(token, ParenExpression):
+        return any([_is_on_demand_search_filter(c) for c in token.children])
```

Comment:
Just a nit, you could rework the recursive implementation of `_contains_on_demand_search_filters` by splitting the function in two:
- `_query_has_on_demand()` -> which parses the query and calls `_contains_on_demand_search_filters` which just loops over the tokens.
- `_contains_on_demand_search_filters` -> which loops over a list of tokens.

Then you could have this like:
```suggestion
        return _contains_on_demand_search_filters(token.children)
```

Reply:
With this you will not implement the same logic twice. What do you think?

Reply:
Yep, good idea. Done in [b0f91a2](https://github.com/getsentry/sentry/pull/52612/commits/b0f91a261a7b266dea4982002ed38ab36c6b3ada)